rfcs/0174-improve-terminology.md

```diff
@@ -0,0 +1,160 @@
+- Nix store outputs that aren't exposed in Nixpkgs are called "derivations". Examples include store paths passed to `src`, symlink farms and other kinds of store paths used in implementation, but not exposed to the end user
+
+In summary, "derivation" concept is an implementation detail. The term, however, is applied to many user-facing scenarios. As a result, the term becomes overloaded, confusing and difficult to comprehend. We should treat derivations as an implementation detail.
+
+Continuing on the topic of derivations, it is clear that store path outputs aren't always directly produced by a derivation. It is not very useful to think in "derivation -> output" chains. Instead, we can use a term "installable" that is [already established](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix#installables) to simplify the terminology:
+
+- `nix build` is a command that converts an installable to an output store path or fails. It is a simple "input -> output" relationship
+- In this model, what is written to the store (aside from the output) are just *dependencies*. This includes plain dependencies (like software libraries), FODs, implementation-based store paths and .drv files
+- Installables are usually packages or runnable scripts. That makes them easy to conceptualize, and might give additional intuition into the future Flake transition. The current unstable schema for Flakes uses installables for all outputs. That said, Flakes also allow definining custom outputs that can be arbitrary
+
+With all that reasoning, here are the actions proposed:
+
+- Rename IFD (Import From Derivation) to IFSP (Import From Store Path)
+- Rename FOD (Fixed-Output Derivation) to WKSP (Well-Known Store Path)
+- Converge on using "package" term for introductory materials
+- Converge on using "installable" term for more advanced materials. This includes cases when an installable is a runnable script, such as when building a closure of NixOS system. The term should also be used in manuals and references
+- Rename "derivation" to "derivation file", keeping the current definition. Define "derivation" as "an old term that is no longer used, which usually means derivation files"
+- Treat every other usage of the word "derivation" as **incorrect documentation**
+- Remove the word "instantiate". For all intents and purposes, "instantiation" is just putting a derivation file into Nix store; this is an implementation detail
+- Replace the words "realise", "realisation" with "install", "installation". Nix operates on installables, and it installs them; there's no need for an arcane term
+- Rename `lib.mkDerivation` function to `lib.mkPackage`. While mkDerivation can be used for defining FOD or IFD, it usually isn't, so we dodge the edge cases where "derivation is not technically a package"
+- Rename `builtins.derivation` to `builtins.drvFile`
```

Comment:
This should probably be specified explicitly that what I meant was "rename X to Y and leave X as an alias to Y". So no backwards-incompatible changes that way, but an old name is clearly deprecated. Good catch.

About renames in general - I suppose renaming `builtins.derivation` doesn't pose a huge problem, yes? For `lib.mkDerivation` - I think a function that is so essential to the whole Nixpkgs and sees so much mention in manuals can be justifiably renamed treewide. This is up in the air and subject to debate, so I'll add that to the "potential considerations" or "drawbacks" section.

Reply:
`builtins.derivation` has a much smaller impact, for obvious reasons (everybody uses `mkDerivation`). `rg "builtins.derivation" --stats` shows just one line. Searching for "derivation" instead matches over 1200, but almost all of those are not the function but strings and comments. I'm not a Nix contributor though, so I can't really comment too deeply on the acceptability of adding a builtin (or renaming, in this case). RFCs 148 and 137 serve as pretty clear examples of how hard it can be to push through Nix language changes though.

Or did you mean that `builtins.derivation` is less significant and therefore doesn't merit renaming? If someone were to inspect a renamed `mkDerivation` and discover that it doesn't make a... whatever we rename it to but instead makes a derivation, how do we resolve that dissonance? I don't think we can, we should rename both or neither.